services/src/main/java/org/keycloak/services/ParValidationService.java

```diff
@@ -0,0 +1,159 @@
+package org.keycloak.services;
+
+import org.jboss.resteasy.spi.HttpRequest;
+import org.keycloak.OAuth2Constants;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequest;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequestParserProcessor;
+import org.keycloak.protocol.oidc.utils.OIDCRedirectUriBuilder;
+import org.keycloak.protocol.oidc.utils.OIDCResponseMode;
+import org.keycloak.protocol.oidc.utils.OIDCResponseType;
+import org.keycloak.protocol.oidc.utils.RedirectUtils;
+import org.keycloak.representations.idm.OAuth2ErrorRepresentation;
+import org.keycloak.util.TokenUtil;
+
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+public class ParValidationService {
+
+    private KeycloakSession session;
+    private EventBuilder event;
+    private AuthorizationEndpointRequest authorizationRequest;
+    private OIDCResponseMode parsedResponseMode;
+    private String redirectUri;
+
+    public ParValidationService(KeycloakSession session, EventBuilder event) {
+        this.session = session;
+        this.event = event;
+    }
+
+    public Response validateParRequest(HttpRequest request, ClientModel clientModel) {
+
+        authorizationRequest = AuthorizationEndpointRequestParserProcessor.parseRequest(event, session, clientModel, request.getFormParameters());
+
+        checkRedirectUri(clientModel);
+        Response errorResponse = checkResponseType(clientModel);
+        if (errorResponse != null) {
+            return errorResponse;
+        }
+
+        if (authorizationRequest.getInvalidRequestMessage() != null) {
+            event.error(Errors.INVALID_REQUEST);
+            return redirectErrorToClient(parsedResponseMode, Errors.INVALID_REQUEST, authorizationRequest.getInvalidRequestMessage());
```

Comment:
I think that error codes might not be implemented properly. This method is for Authorization Endpoint. PAR does not relate to HTTP redirect.